rust/lance-index/src/scalar/ngram.rs

```diff
@@ -0,0 +1,456 @@
+        _dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        todo!()
+    }
+
+    /// Add the new data into the index, creating an updated version of the index in `dest_store`
+    async fn update(
+        &self,
+        _new_data: SendableRecordBatchStream,
+        _dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        todo!()
+    }
+}
+
+pub struct NGramIndexBuilder {
+    tokenizer: TextAnalyzer,
+    tokens_map: HashMap<String, u32>,
+    tokens_counter: u32,
+    bitmaps: Vec<RoaringTreemap>,
+}
+
+impl Default for NGramIndexBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl NGramIndexBuilder {
+    pub fn new() -> Self {
+        let tokenizer = NGRAM_TOKENIZER.clone();
+        Self {
+            tokenizer,
+            // Default capacity loosely based on case insensitive ascii ngrams with punctuation stripped
+            tokens_map: HashMap::with_capacity(4 * 26 * 26 * 26),
+            bitmaps: Vec::with_capacity(4 * 26 * 26 * 26),
+            tokens_counter: 0,
+        }
+    }
+
+    fn validate_schema(schema: &Schema) -> Result<()> {
+        if schema.fields().len() != 2 {
+            return Err(Error::InvalidInput {
+                source: "Ngram index schema must have exactly two fields".into(),
+                location: location!(),
+            });
+        }
+        if *schema.field(0).data_type() != DataType::Utf8 {
+            return Err(Error::InvalidInput {
+                source: "First field in ngram index schema must be of type Utf8".into(),
+                location: location!(),
+            });
+        }
+        if *schema.field(1).data_type() != DataType::UInt64 {
+            return Err(Error::InvalidInput {
+                source: "Second field in ngram index schema must be of type UInt64".into(),
+                location: location!(),
+            });
+        }
+        Ok(())
+    }
+
+    fn process_batch(&mut self, batch: &RecordBatch) {
+        let text_col = batch.column(0).as_string::<i32>();
+        let row_id_col = batch.column(1).as_primitive::<UInt64Type>();
+        for (text, row_id) in text_col.iter().zip(row_id_col.values()) {
+            let text = text.unwrap();
+            let tokens = self.tokenizer.clone().tokenize(text);
+            for token in tokens {
+                let token_id = *self.tokens_map.entry(token).or_insert_with(|| {
+                    let token_id = self.tokens_counter;
+                    self.tokens_counter += 1;
+                    self.bitmaps.push(RoaringTreemap::new());
+                    token_id
+                });
+                self.bitmaps[token_id as usize].insert(*row_id);
+            }
+        }
+    }
+
+    pub async fn train(&mut self, mut data: SendableRecordBatchStream) -> Result<()> {
+        let schema = data.schema();
+        Self::validate_schema(schema.as_ref())?;
+
+        while let Some(batch) = data.try_next().await? {
+            self.process_batch(&batch);
```

Comment:
how easy would this be to parallelize? would this be the place in the code to do that?

Reply:
It would be good to parallelize.  Perhaps not trivially easy because we have to build up a series of per-thread maps and then merge them in the end.  We do this in the trainer for the inverted index.  We also spill that index to disk but I don't think we will need to do that here since the # of trigrams is bounded.

Reply:
Deferring for follow-up to keep this PR constrained for now.